imap_processing/ultra/l2/ultra_l2.py

```diff
@@ -148,11 +167,143 @@ def get_variable_attributes_optional_energy_dependence(
+    if bin_edges[-1] != n_fine_energy_bins:
+        bin_edges = np.append(bin_edges, n_fine_energy_bins)
+
+    return bin_edges
+
+
+def bin_pset_energy_bins(
+    pset: xr.Dataset, bin_groups: np.ndarray | None = None
+) -> xr.Dataset:
+    """
+    Group fine-grained L1C PSET energy bins into coarser bins for l2 ULTRA maps.
+
+    Parameters
+    ----------
+    pset : xarray.Dataset
+        Ultra L1C pointing set dataset to bin.
+    bin_groups : numpy.ndarray, optional
+        Array of indices defining the new energy bin edges. If not provided,
+        DEFAULT_BIN_EDGES will be used.
+
+    Returns
+    -------
+    xarray.Dataset
+        The input pset with energy bins grouped according to the bin_groups.
+    """
+    if bin_groups is None:
+        bin_groups = build_default_coarse_bin_edges()
+    # Get a list of variables that have the energy bin dimension
+    energy_dep_vars = [
+        var for var in pset.data_vars if "energy_bin_geometric_mean" in pset[var].dims
+    ]
+    # From those, get the variables that need to be averaged over the coarse bins
+    vars_to_average = [
+        var
+        for var in energy_dep_vars
+        if var in VARIABLES_TO_AVERAGE_OVER_COARSE_ENERGY_BINS
+    ]
+    n_fine_bins = pset["energy_bin_geometric_mean"].size
+    logger.info(
+        f"Binning pset fine energy bins into coarser bins with edges: {bin_groups}"
+    )
+
+    if bin_groups[-1] > n_fine_bins:
+        raise ValueError(
+            "The given bin_groups contain an index larger than the number of "
+            f"fine energy bins in the pset: {n_fine_bins}."
+        )
+    # Add any missing 'catch-all' bins at the start or end
+    if bin_groups[-1] < n_fine_bins:
+        bin_groups = np.append(bin_groups, n_fine_bins)
+        logger.info(
+            "The given bin_groups did not include the last fine energy bins."
+            f" Adding another 'catch-all' bin. New coarse energy bin edges:"
+            f" {bin_groups}"
+        )
+    if bin_groups[0] > 0:
+        bin_groups = np.insert(bin_groups, 0, 0)
+        logger.info(
+            "The given bin_groups did not include the first fine energy bins."
+            f" Adding another 'catch-all' bin. New coarse energy bin edges:"
+            f" {bin_groups}"
+        )
+    # Create a new coordinate for the new energy bin index
+    # For example, if bin_groups = [0,4,8,12...46], then the new coordinate will be:
+    # energy_bin_index = [0,0,0,0,1,1,1,1,2,2,2,2...12] That way we can groupby the new
+    # energy bin index to sum/average over the fine bins.
+    pset = pset.assign_coords(
+        energy_bin_index=(
+            "energy_bin_geometric_mean",
+            np.digitize(np.arange(n_fine_bins), bin_groups, right=False),
+        )
+    )
+    # Count number of pixels
+    non_zero_pixels_per_group = (
+        ((pset[vars_to_average] != 0) | (pset[vars_to_average] != FILLVAL_FLOAT32))
```

Comment:
and? Otherwise your FILLVAL_FLOAT32 will match the first condition here, but you do want to ignore it.